core/domain/state_domain.py

```diff
@@ -540,6 +579,85 @@ def from_dict(cls, interaction_dict):
+        elif self.id == 'MultipleChoiceInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                multiple_choice_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.hints, self.answer_groups))
+        elif self.id == 'NumericInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                numeric_input=interaction_instance.to_proto(
+                    self.default_outcome, self.solution,
+                    self.hints, self.answer_groups))
+        elif self.id == 'TextInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                text_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.solution, self.hints, self.answer_groups))
+        elif self.id == 'RatioExpressionInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                ratio_expression_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.solution, self.hints, self.answer_groups))
+        elif self.id == 'ImageClickInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                image_click_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.hints, self.answer_groups))
+        elif self.id == 'DragAndDropSortInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                drag_and_drop_sort_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.solution, self.hints, self.answer_groups))
+        elif self.id == 'AlgebraicExpressionInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                algebraic_expression_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.solution, self.hints, self.answer_groups))
+        elif self.id == 'MathEquationInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                math_equation_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.solution, self.hints, self.answer_groups))
+        elif self.id == 'NumericExpressionInput':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                numeric_expression_input=interaction_instance.to_proto(
+                    self.default_outcome, self.customization_args,
+                    self.solution, self.hints, self.answer_groups))
+        elif self.id == 'EndExploration':
+            interaction_proto = state_pb2.InteractionInstanceDto(
+                end_exploration=interaction_instance.to_proto())
```

Comment:
I am not inclined towards raising exception here, if the code comes to the else part then it is ok to just return the empty dict.
```
interaction_proto = {}
```

Reply:
Have you confirmed this with @BenHenning? This is an important part of the specification so we cannot make arbitrary decisions here. Either we do something like UnknownInteraction (similar for languages) or we raise an error here because such an exploration should not be convertable into proto and we shouldn't have tried to do it in the first place -- so, this is a question about the requirements.

Reply:
I think there are two different desired outcomes that are slightly warring with each other here:
1. The fact that we don't want things like empty protos filled in for the Android app because we should only ever be downloading something that's guaranteed to be supported
2. The fact that we need to be able to convert non-supported explorations to proto when computing sizes

I think failing here is fine (and better than doing a default to fulfill (1)), but throwing an exception seems incorrect per (2) because it's an okay failure for unsupported explorations. Is there another error mechanism that we can use in this case?